sympy/polys/matrices/tests/test_fflu.py

```diff
@@ -0,0 +1,299 @@
+    ),
+
+    (
+        'zz_nx1_2rows',
+        DM([[81], [54]], ZZ),
+        DM([[1, 0], [0, 1]], ZZ),
+        DM([[81, 0], [54, 81]], ZZ),
+        DM([[81, 0], [0, 81]], ZZ),
+        DM([[81], [0]], ZZ),
+    ),
+
+    (
+        'zz_nx2_3rows',
+        DM([[2, 7], [7, 45], [25, 84]], ZZ),
+        DM([[1, 0, 0], [0, 1, 0], [0, 0, 1]], ZZ),
+        DM([[2, 0, 0], [7, 82, 0], [25, 41, 41]], ZZ),
+        DM([[2, 0, 0], [0, 82, 0], [0, 0, 41]], ZZ),
+        DM([[2, 7], [0, 82], [0, 0]], ZZ),
+    ),
+
+    (
+
+        'zz_1x2',
+        DM([[0, 28]], ZZ),
+        DM([[1]], ZZ),
+        DM([[28]], ZZ),
+        DM([[28]], ZZ),
+        DM([[0, 28]], ZZ)
+    ),
+
+    (
+        'zz_nx3_4rows',
+        DM([[84, 30, 9], [20, 59, 13], [53, 46, 81], [63, 48, 29]], ZZ),
+        DM([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]], ZZ),
+        DM([[84, 0, 0, 0], [20, 365904, 0, 0], [53, 303411, 303411, 0], [63, 303411, 303411, 303411]], ZZ),
+        DM([[84, 0, 0, 0], [0, 365904, 0, 0], [0, 0, 1321658316, 0], [0, 0, 0, 303411]], ZZ),
+        DM([[84, 30, 9], [0, 365904, 13], [0, 0, 1321658316], [0, 0, 0]], ZZ),
+    ),
+]
+
+
+def _check_fflu(A, P, L, D, U):
+    A_field = _to_DM(A, P).to_field().to_dense()
+    P_field = P.to_field().to_dense()
+    L_field = L.to_field().to_dense()
+    D_field = D.to_field().to_dense()
+    U_field = U.to_field().to_dense()
+    m, n = A.shape
+    assert P_field.shape == (m, m)
+    assert L_field.shape == (m, m)
+    assert D_field.shape == (m, m)
+    assert U_field.shape == (m, n)
+    assert L_field.is_lower
+    assert D_field.is_diagonal
+    if hasattr(D, 'inv_den'):
+        di, d = D.inv_den()
+        assert P.matmul(A).rmul(d) == L.matmul(di).matmul(U)
+    else:
+        assert L_field.matmul(D_field).matmul(U_field) == P_field.matmul(A_field)
+    assert U_field.is_upper
+
+
+def _to_DM(A, ans):
+    if isinstance(A, DomainMatrix):
+        return A
+    elif isinstance(A, Matrix):
+        return A.to_DM(ans.domain)
+
+    if not (hasattr(A, 'shape') and hasattr(A, 'domain')):
+        shape, domain = ans.shape, ans.domain
+    else:
+        shape, domain = A.shape, A.domain
+
+    if isinstance(A, DDM):
+        return DomainMatrix(A.to_list(), shape, domain)
+    elif isinstance(A, SDM):
+        return DomainMatrix(A.to_list(), shape, domain)
+    elif isinstance(A, DFM):
+        data = A.to_ddm().to_list()
+        return DomainMatrix(data, shape, domain)
+    else:
+        raise TypeError(f"Cannot convert {type(A)} to DomainMatrix")
+
+
+def _check_fflu_result(result, A, P_ans, L_ans, D_ans, U_ans):
+    P, L, D, U = result
+    P = _to_DM(P, P_ans)
+    L = _to_DM(L, L_ans)
+    D = _to_DM(D, D_ans)
+    U = _to_DM(U, U_ans)
+    A = _to_DM(A, P_ans)
+    m, n = A.shape
+    assert P.shape == (m, m)
+    assert L.shape == (m, m)
+    assert D.shape == (m, m)
+    assert U.shape == (m, n)
+    assert L.is_lower
+    assert D.is_diagonal
+    if hasattr(D, 'inv_den'):
```

Comment:
I think that is should always have the `inv_den` method because `D` is a DomainMatrix.